bsms.mediawiki

```diff
@@ -0,0 +1,169 @@
+* Abort the setup if there are any <tt>KEY</tt> duplicates.
+* The Signer checks whether one of the <tt>KEY</tt>s in the descriptor belongs to it, using path and fingerprint information included in the descriptor. The check must perform an exact match on the <tt>KEY</tt>s, and not using shortcuts such as matching fingerprints (which is trivial to spoof). Abort the setup if it doesn’t detect its own <tt>KEY</tt>.
+* For confirmation, the Signer must display to the user the <tt>CHECKSUM</tt>, plus important policy parameters, such as <tt>M</tt>, <tt>N</tt> and the position of its own XPUB in the policy script. The total number of Signers, <tt>N</tt>, is important to prevent a <tt>KEY</tt> insertion attack. The position is important for scripts where key order matters. Except for the position, all participating Signers should be able to display the same confirmation.
+* If all checks pass, the Signer must persist the descriptor record in its storage.
+* The Signer can choose to further restrict post-XPUB derivation paths, such as to those defined in [https://github.com/bitcoin/bips/blob/master/bip-0044.mediawiki BIP-0044].
+* The Signer should subsequently use the descriptor to generate and verify receive and change addresses.
+
+This completes the setup.
+
+===Encryption===
+
+====The Token====
+We define three modes of encryption.
+
+# <tt>NO_ENCRYPTION</tt> : the <tt>TOKEN</tt> is set to <tt>0</tt>. Encryption is disabled.
+# <tt>STANDARD</tt> : the <tt>TOKEN</tt> is a 64-bit nonce.
+# <tt>EXTENDED</tt> : the <tt>TOKEN</tt> is a 96-bit nonce.
+
+The <tt>TOKEN</tt> can be converted to one of these formats:
+* A mnemonic phrase using [https://github.com/bitcoin/bips/blob/master/bip-0039.mediawiki BIP-0039] word list (6 words in <tt>STANDARD</tt> mode, 9 words in <tt>EXTENDED</tt> mode).
+* A decimal number. The number must not exceed 2^64 in <tt>STANDARD</tt> mode, and 2^96 in <tt>EXTENDED</tt> mode.
+* A QR code.
+* Other formats.
+
+The flexibility in the data format allows each Signer to customize the User Experience based on its respective capabilities.
+
+====Key Derivation====
+The key derivation function is [https://tools.ietf.org/html/rfc2898 PBKDF2], with PRF = SHA512. Specifically:
+
+<tt>DK = PBKDF2(PRF, Password, Salt, c, dkLen)</tt>
+
+Whereas:
+
+* PRF = <tt>SHA512</tt>
+* Password = <tt>"No SPOF"</tt>
+* Salt = <tt>TOKEN</tt>
+* c = <tt>2048</tt>
+* dkLen = <tt>256</tt>
+* DK = Derived <tt>ENCRYPTION_KEY</tt>
+
+====Encryption Scheme====
+The encryption scheme is [https://tools.ietf.org/html/rfc3686 AES-256-CTR].
+
+<tt>Ciphertext = AES-256-CTR-Encrypt(Plaintext, Key, IV)</tt>
+
+<tt>Plaintext = AES-256-CTR-Decrypt(Ciphertext, Key, IV)</tt>
+
+Whereas:
+* Key = <tt>ENCRYPTION_KEY</tt>, derived from the <tt>TOKEN</tt>
+* IV = <tt>SHA256(ENCRYPTION_KEY + K)</tt>, where K is the XPUB in the key record, and the descriptor checksum in the descriptor record
```

Comment:
What is the motivation for choosing IV like this? AFAIK IV is required to be unique for the given key, but otherwise can be any value. Since there's only two packets, couldn't IV be simple 0 and 1? There won't be the need to send IV along with the packet then.

Reply:
I thought about hardcoding it to 0 and 1. But for Signer->Coordinator that means we will reuse the IV among the Signers (if a common TOKEN is shared between them).

Reply:
I see, with Signer->Coordinator the fixed IV cannot be used then. I think it should be specified if a binary representation of the xpub, or the base58-encoded string representation should be used in generating IV.

With Coordinator->Signer, there's only a single descriptor record sent to all signers, so it shouldn't be a problem to use the fixed IV. Might make things a bit simpler at least on coordinator side. It might make sense to use similar procedures to generate IV for both directions, though, just for uniformity.

Reply:
> With Coordinator->Signer, there's only a single descriptor record sent to all signers, so it shouldn't be a problem to use the fixed IV

That would simplify things, but there's a (extremely low) chance of collision?

Reply:
> It might make sense to use similar procedures to generate IV for both directions, though, just for uniformity.

Yeah, pro/con.

Reply:
If there's some misconfiguration/error that results in two signers using the same xpub, they might reuse the same `SHA256(ENCRYPTION_KEY + XPUB)` to encrypt different records. It could be a mere error that can be fixed, but now the session will need to be restarted with new token.  Maybe use the signature of the key record instead ? It covers all the lines of the key record, so should be unique regardless of xpub.

Reply:
>> With Coordinator->Signer, there's only a single descriptor record sent to all signers, so it shouldn't be a problem to use the fixed IV

> That would simplify things, but there's a (extremely low) chance of collision?

The collision in TOKENs ? Then I doubt a hash-derived IV will make any difference

Reply:
>  Maybe use the signature of the key record instead ?

or just the hash of the entire key record (it will need to be computed to make a signature, anyway)

Reply:
> If there's some misconfiguration/error that results in two signers using the same xpub, they might reuse the same SHA256(ENCRYPTION_KEY + XPUB) to encrypt different records. It could be a mere error that can be fixed, but now the session will need to be restarted with new token. Maybe use the signature of the key record instead ? It covers all the lines of the key record, so should be unique regardless of xpub.

That (two signers using the same XPUB) is unlikely IMO. Otoh, switching to using the signature wouldn't hurt.

Reply:
>>> With Coordinator->Signer, there's only a single descriptor record sent to all signers, so it shouldn't be a problem to use the fixed IV

>> That would simplify things, but there's a (extremely low) chance of collision?

> The collision in TOKENs ? Then I doubt a hash-derived IV will make any difference

I mean the collision in IVs. If we use fixed IV for Coordinator->Signer, there's a chance that the IV the Signers derive will collide with whatever fixed IV we come up with. But I also see the same problem whether we use a fixed IV or not. Given how unlikely it is though, probably not a big deal.

An additional advantage of a hash-derived IV approach in both directions is that we can potentially reuse the session key (say if you want to set up multiple multisig wallets).

Reply:
> If we use fixed IV for Coordinator->Signer, there's a chance that the IV the Signers derive will collide with whatever fixed IV we come up with. 

If the IV is fixed, it is trivial to detect the collision, though, and abort. Also, SHA256 collision will break a lot more systems than just this :-)

> An additional advantage of a hash-derived IV approach in both directions is that we can potentially reuse the TOKEN (say if you want to set up multiple multisig wallets), since all packets will have unique IV.

Agreed.

But if we're talking about collisions, the probability of collision on descriptor checksum is much bigger than SHA256 collision. If you reuse the same token for many multisig setups, it becomes theoretically possible (although unlikely that it is practically possible)

Reply:
> But if we're talking about collisions, the probability of collision on descriptor checksum is much bigger than SHA256 collision, since checksum is designed for different purpose

We can switch to descriptor instead of descriptor checksum.

Reply:
Maybe just use `SHA256(TOKEN + <whole_record>)` for IV in both cases ?

Reply:
that could work.

Reply:
I just realize that if we HMAC-SHA() over the entire record, the IV could double as the MAC. So we do have authentication :-)

Reply:
Quick summary of off-thread conversation with @dgpv:

If we use `HMAC-SHA512(TOKEN + <whole_record>)`* as the IV, the IV can actually double as the MAC.

Basically, this would be the Encrypt-and-MAC scheme as described here https://en.wikipedia.org/wiki/Authenticated_encryption#Encrypt-and-MAC_(E&M)

Between this and the built-in checksums of XPUB/descriptor, plus the signature from Signer->Coordinator, this could offer enough integrity protection (TBD).

* `HMAC-SHA512` because it is already available on most HWWs due to BIP32.

Reply:
Updated IV calculation: https://github.com/nunchuk-io/bips/pull/1/commits/71d7745df79180a7de80446ef1b274523fdd3271

Reply:
IV verification: https://github.com/nunchuk-io/bips/pull/1/commits/709d5886e001f0d0137d0e27fde51df295a22274